A6.md

```diff
@@ -0,0 +1,398 @@
+        * [Disabling Retries](#disabling-retries)
+        * [Retry and Hedging Statistics](#retry-and-hedging-statistics)
+     * [Configuration Language](#configuration-language)
+        * [Retry Policy](#retry-policy)
+        * [Hedging Policy](#hedging-policy-1)
+        * [Throttling Configuration](#throttling-configuration)
+        * [Integration with Service Config](#integration-with-service-config)
+  * [Open issues](#open-issues)
+     * [Security](#security)
+
+## Abstract
+
+gRPC client library will automatically retry failed RPCs according to a policy set by the service owner.
+
+## Background
+
+Currently, gRPC does not retry failed RPCs. All failed RPCs are immediately returned to the application layer by the gRPC client library.
+
+Many teams have implemented their own retry logic wrapped around gRPC like [Veneer Toolkit](https://github.com/googleapis/toolkit) and [Cloud Bigtable](https://github.com/GoogleCloudPlatform/cloud-bigtable-client).
+
+## Proposal
+
+### Overview
+
+gRPC will support two configurable retry policies. The service configuration may choose from a retry policy (retry failed RPCs) or a hedging policy (aggressively send the same RPC multiple times in parallel). An individual RPC may be governed by a retry policy or a hedge policy, but not both.
+
+Retry policy capabilities are as follows. Each has a detailed description below.
+* [Maximum number of retry attempts](#maximum-number-of-retries)
+* [Exponential backoff](#exponential-backoff)
+* [Set of retryable status codes](#retryable-status-codes)
+
+The hedging policy has the following parameters. See details [here](#hedging-policy).
+* Maximum number of hedged requests
+* Delay between hedged requests
+* Set of non-fatal status codes
+
+Additionally, gRPC provides a mechanism to throttle retry attempts and hedged RPCs when the ratio of failures to successes exceeds a threshold. See [detailed description of throttling](#throttling-retry-attempts-and-hedged-rpcs).
+
+We also provide a mechanism for servers to explicitly signal clients to retry after a settable delay. See [detailed description of server pushback](#pushback).
+
+In some cases, gRPC can guarantee that a request has never been seen by the server application logic. These cases will be transparently retried by gRPC, as detailed [here](#transparent-retries).
+
+Lastly, information about number of retry attempts will be exposed to the client and server applications through metadata. Find more details [here](#exposed-retry-metadata).
+
+### Detailed Design
+
+#### Retry Policy Capabilities
+
+Retry policies support configuring the maximum number of retries, the parameters for exponential backoff, and the set of retryable status codes. Each of these configuration options is detailed in its own section below.
+
+##### Maximum Number of Retries
+
+The maximum number of retry attempts per original RPC. This can be specified in the retry policy as follows:
+
+```
+'max_retry_attempts': 3
```

Comment:
The example for the hedging policy below shows the parameters wrapped in a 'hedging_policy' block.  For consistency, the retry examples should be wrapped in a 'retry_policy' block.